src/compiler/msvc.rs

```diff
@@ -286,12 +301,58 @@ pub fn parse_arguments(arguments: &[OsString]) -> CompilerArguments<ParsedArgume
             }
         }
         match item.data {
-            Some(PreprocessorArgument) |
+            Some(PreprocessorArgument) => preprocessor_args.extend(item.arg.normalize(NormalizedDisposition::Concatenated)),
             Some(ProgramDatabase) |
             Some(DebugInfo) => common_args.extend(item.arg.normalize(NormalizedDisposition::Concatenated)),
             _ => {}
         }
     }
+
+    // TODO: doing this here reorders the arguments, hopefully that doesn't affect the meaning
+    let xclang_it = gcc::ExpandIncludeFile::new(cwd, &xclangs);
+    for item in ArgsIter::new(xclang_it, (&gcc::ARGS[..], &clang::ARGS[..])) {
+        // Eagerly bail if it looks like we need to do more complicated work
+        use compiler::gcc::GCCArgAttribute::*;
+        let args = match item.data {
+            Some(SplitDwarf) |
+            Some(ProfileGenerate) |
+            Some(TestCoverage) |
+            Some(Coverage) |
+            Some(DoCompilation) |
+            Some(Language) |
+            Some(Output) |
+            Some(TooHard) => {
+                return CompilerArguments::CannotCache(item.arg.to_str().unwrap_or(
+                    "Can't handle complex arguments through clang",
+                ))
+            }
+            None => {
+                match item.arg {
+                    Argument::Raw(_) |
+                    Argument::UnknownFlag(_) => Some(&mut common_args),
+                    _ => unreachable!(),
+                }
+            }
+            Some(PassThrough) => Some(&mut common_args),
+            Some(PreprocessorArgument) |
+            Some(DepTarget) |
+            Some(NeedDepTarget) => Some(&mut preprocessor_args),
+        };
+        if let Some(args) = args {
+            // Normalize attributes such as "-I foo", "-D FOO=bar", as
+            // "-Ifoo", "-DFOO=bar", etc. and "-includefoo", "idirafterbar" as
+            // "-include foo", "-idirafter bar", etc.
+            let norm = match item.arg.to_str() {
+                Some(s) if s.len() == 2 => NormalizedDisposition::Concatenated,
+                _ => NormalizedDisposition::Separated,
+            };
+            for arg in item.arg.normalize(norm) {
+                args.push("-Xclang".into());
+                args.push(arg)
+            }
+        }
+    }
+
```

Comment:
For the purposes of distributed compilation it's not sufficient to just 'pass through' any `-Xclang` options as suggested in #262 - it's important to filter out the preprocessor-only args, which the dependency stuff falls under.

I'm not sure what the best solution long-term is - by the time you've created a clang (or rather, gcc) `ParseArguments` you've lost a lot of information (e.g. "there is another input file" turns into "here's an arg") which reduces our accuracy. I think you almost want an interim `LosslessParsedArgs` so you can inspect it e.g. within this function to see how it affects the msvc compilation.

For now it just grabs the pieces that can be easily partitioned into common or preprocessor args without any side effects and bails otherwise.

Reply:
This seems like a reasonable way to handle things for now. It might be worth refactoring things in the future so that we could use something less leaky like `parse_arguments`, but I won't lose any sleep over it. clang-cl is an odd duck, and dealing with `-Xclang` is even odder. I'd probably want to read the clang-cl source to figure things out before anyone spent any more time on this.